### PR TITLE
Issue #22: Handling statistics for weapons

### PR DESCRIPTION
**Pull request by belmirp**
The feature works as follows:

- statistics are posted to endpoint: **/ut/api/stats/accountId/{accountId}/bulk?ownertype=1** after each match
- posted statistics are saved with **window=daily** flag
- you can get current account's statistics through **/ut/api/stats/accountId/{accountId}/bulk/window/{window}** endpoint
- daily/weekly/monthly statistics are calculated on-the-fly each time a request is made to GET endpoints
- all-time statistics are calculated each time a new daily statistic is added and saved to the DB, so when requested they are fetched from the DB

image

### CODE REVIEW

File: UT4MasterServer/Services/DatabaseContext.cs
Comment (timiimit) [inline UT4MasterServer/Services/DatabaseContext.cs:23]: I'm not familiar with indexes... is it intentional to create them upon EVERY access to database? new instance of DatabaseContext is created on each request. what does this exactly do? i think this would be better suited to be placed in StatisticsService.

File: UT4MasterServer/Services/StatisticsService.cs
Comment (timiimit) [inline UT4MasterServer/Services/StatisticsService.cs:16]: `allowPasswordGrant` is never used. it can be removed.

File: UT4MasterServer/Services/StatisticsService.cs
Comment (timiimit) [inline UT4MasterServer/Services/StatisticsService.cs:28]: just like everywhere else `accountId` should be of type `EpicID`.

im not sure if this method is used, but two of them below look like copy-paste. cant they my merged into one?
`GetAggregateAccountStatistics` already seems like it can be used for any windows of statistics, no?

File: UT4MasterServer/Services/StatisticsService.cs
Comment (timiimit) [inline UT4MasterServer/Services/StatisticsService.cs:139]: Remove spaces between calls

File: UT4MasterServer/Services/StatisticsService.cs
Comment (timiimit) [inline UT4MasterServer/Services/StatisticsService.cs:2124]: This copy paste is alright for now, but this definitely needs some improvement in the future...

File: UT4MasterServer/Enums/StatisticWindow.cs
Comment (timiimit) [inline UT4MasterServer/Enums/StatisticWindow.cs:3]:
```suggestion
public enum StatisticWindow : int
```
if this is being used as an integer representing number of days, it could inherit from int

**Review by timiimit — CHANGES_REQUESTED** [review]
Looks alright overall, but i would perfer some changes before merging. I left comments on the code.

Account ids should always be `EpicID` type.

File: UT4MasterServer/Controllers/StatisticsController.cs
Comment (belmirp) [reply UT4MasterServer/Controllers/StatisticsController.cs:14]: Didn't see it, I will move it there.

File: UT4MasterServer/DTOs/StatisticBulkDto.cs
Comment (belmirp) [reply UT4MasterServer/DTOs/StatisticBulkDto.cs:4]: DTO stands for Data Transfer Object. It is like ViewModel but more generic. Models should represent only database tables/fields. DTOs should represent requests or responses, or any data that is passed throughout the application, so only needed fields, not the whole model. You should decide whether to use DTOs or ViewModels, in my opinion DTO is more appropriate.

File: UT4MasterServer/DTOs/StatisticDto.cs
Comment (belmirp) [reply UT4MasterServer/DTOs/StatisticDto.cs:6]: Same response as above.

File: UT4MasterServer/Models/Statistic.cs
Comment (belmirp) [reply UT4MasterServer/Models/Statistic.cs:14]: I will try to replace string with EpidID type.

File: UT4MasterServer/Services/StatisticsService.cs
Comment (belmirp) [reply UT4MasterServer/Services/StatisticsService.cs:28]: The methods were different, but I changed something. I will merge them into one.

File: UT4MasterServer/Services/StatisticsService.cs
Comment (belmirp) [reply UT4MasterServer/Services/StatisticsService.cs:2124]: It is possible to do it in one foreach loop using reflection, but that is way slower. I don't recommend it.

File: UT4MasterServer/Enums/StatisticWindow.cs
Comment (belmirp) [reply UT4MasterServer/Enums/StatisticWindow.cs:3]: Every enum is of int type by default, no need to mention it.

File: UT4MasterServer/DTOs/StatisticBulkDto.cs
Comment (timiimit) [reply UT4MasterServer/DTOs/StatisticBulkDto.cs:4]: alright, lets leave the naming as is

File: UT4MasterServer/Models/Statistic.cs
Comment (timiimit) [reply UT4MasterServer/Models/Statistic.cs:14]: EpicID is automatically converted to strings when serialized to bson. you can take a look at other model classes for example

File: UT4MasterServer/Services/StatisticsService.cs
Comment (timiimit) [reply UT4MasterServer/Services/StatisticsService.cs:2124]: Yea, definitely don't want to use reflection. I just thought i'd comment about it.

File: UT4MasterServer/Controllers/UnrealTournamentStatsController.cs
Comment (timiimit) [inline UT4MasterServer/Controllers/UnrealTournamentStatsController.cs:33]: Leave comment that you didn't handle
```suggestion
	// TODO: Check id and return ErrorResponse if invalid: "Could not convert fakeAccountId to a UUID because it was not in a valid format: Invalid UUID string: fakeAccountId" (errors.com.epicgames.modules.stats.invalid_account_id)
	var accountId = EpicID.FromString(id);
```

File: UT4MasterServer/Models/Statistic.cs
Comment (belmirp) [reply UT4MasterServer/Models/Statistic.cs:12]: Sure. This is out of habit. There are some static analyzers that warn you to use PascalCase everywhere, so ID would be Id, DTO would be Dto, etc.
I will change it to ID.

File: UT4MasterServer/Services/DatabaseContext.cs
Comment (belmirp) [reply UT4MasterServer/Services/DatabaseContext.cs:23]: The command is checking if the index with same definition and name already exists, so they only created once, but checked every time.
I will see to put this to another place, and run it only once when application starts.

**Review by timiimit — APPROVED** [review]

File: UT4MasterServer/Services/DatabaseContext.cs
Comment (timiimit) [reply UT4MasterServer/Services/DatabaseContext.cs:23]: I already do something like this in CloudStorageService

**Review by timiimit — APPROVED** [review]

**Pull request merged**
